settings: add support for eeprom storage

Laczen opened the pull request:
Add the possibility to the settings subsystem to use eeprom storage. On the eeprom each name-value (id-value) entry is represented as a record containing:
```
------------------------------------------
| reclen | idmaxidx | id  | data | crc16 |
------------------------------------------
with:
	reclen: record length: (uint16_t)
	idmaxidx: length of id - 1: (uint8_t)
	id: id or name
	data: data or value
	crc16: calculated over id + data: (uint16_t)
```

These records are stored one after the other. Records are deleted by making the crc16 invalid. When there is no more place to write a new record a data compression routine is called to remove all deleted records from eeprom storage.

The proposed storage allows to specify a limited region of an eeprom as storage. The default however is to use the full eeprom0 area for storage.

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

henrikbrixandersen reviewed (changes requested):
@Laczen Thank you. Please split the additions to the settings subsystem and the tests into separate commits.

Additionally, it would be nice if this new backend was mentioned in https://docs.zephyrproject.org/latest/reference/settings/index.html#zephyr-storage-backends

Laczen commented:
> @Laczen Thank you. Please split the additions to the settings subsystem and the tests into separate commits.
> 
> Additionally, it would be nice if this new backend was mentioned in https://docs.zephyrproject.org/latest/reference/settings/index.html#zephyr-storage-backends

@brix, I will split up the commit, but for my understanding, what is the reason to do this ?

henrikbrixandersen commented:
> @brix, I will split up the commit, but for my understanding, what is the reason to do this ?

Thanks. I believe it is a well established, good practice to break down changes in logically self-contained commits.
There is a, rather vague, description of this in the Zephyr contribution guidelines: https://docs.zephyrproject.org/latest/contribute/index.html#other-commit-expectations

Laczen committed:
settings: add support for eeprom storage

This adds the possibility to the settings subsystem to use eeprom
storage. On the eeprom each name-value (id-value) entry is represented
as a record containing:
------------------------------------------
| reclen | idmaxidx | id  | data | crc16 |
------------------------------------------
with:
        reclen: record length: (uint16_t)
        idmaxidx: length of id - 1: (uint8_t)
        id: id or name
        data: data or value
        crc16: calculated over id + data: (uint16_t)

These records are stored one after the other. Records are deleted by
making the crc16 invalid. When there is no more place to write a new
record a data compression routine is called to remove all deleted
records from eeprom storage.

The proposed storage allows to specify a limited region of an eeprom as
storage. The default however is to use the full eeprom0 area for
storage.

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

Laczen committed:
tests: add tests for eeprom settings storage

Add the tests to verify the usage for eeprom as a settings backend

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

Laczen committed:
doc: extend settings docs for eeprom storage

Add a fourth possibility for settings storage using an EEPROM as a
backend.

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

Laczen commented:
> > @brix, I will split up the commit, but for my understanding, what is the reason to do this ?
> 
> Thanks. I believe it is a well established, good practice to break down changes in logically self-contained commits.
> There is a, rather vague, description of this in the Zephyr contribution guidelines: https://docs.zephyrproject.org/latest/contribute/index.html#other-commit-expectations

@brix, I have done the split.

However I think the "well established, good practice" is wrong. For me it just adds extra work when changes need to be applied, reorganizing the commits so the changes are added to the correct commit, ...(but maybe this is because my github knowledge is to limited). This extra work is only to get it into "logical self-contained commits" that for me is at the PR level.

nvlsianpu reviewed (commented):
Thanks for the patch, will review soon.

nvlsianpu commented on subsys/settings/src/settings_eeprom.c:
```diff
@@ -0,0 +1,483 @@
+	 */
+	while (1) {
+		rc = settings_eeprom_recstep(cf, &addr, &info, (uint8_t *)id);
+		if (rc <= 0) {
```
I might misunderstand the algorithm, but I think `rc == 0`  for invalidated settings record after which a valid record might be place.

Laczen replied:
@nvlsianpu, thanks for the review. `settings_eeprom_recstep()` only returns 0 at the end of the fs, deleted records are not returned.

de-nordic commented on subsys/settings/src/settings_eeprom.c:
```diff
@@ -0,0 +1,483 @@
+	off_t addr = cf->start + sizeof(struct settings_eeprom_hdr);
+	off_t wraddr = cf->start + sizeof(struct settings_eeprom_hdr);
+	uint16_t reclen;
+	uint8_t buf[4];
```
Why 4 it is 8 in `settings_eeprom_recstep`, why not the same?

de-nordic commented on subsys/settings/src/settings_eeprom.c:
```diff
@@ -0,0 +1,483 @@
+			reclen -= cplen;
+		}
+
+		/* invalidate the record at addr unless it is at wraddr */
```
Maybe "invalidate already copied record unless it is record that has just been written"; it is not at the address, the address is actually pointing after the invalidated record. Comment is misleading.

de-nordic commented on subsys/settings/src/settings_eeprom.c:
```diff
@@ -0,0 +1,483 @@
+		return -EIO;
+	}
+
+	while (1) {
+		rc = settings_eeprom_recstep(cf, &addr, NULL, NULL);
+		if (rc <= 0) {
+			break;
+		}
+		reclen = (uint16_t)rc;
+		addr -= reclen;
+
+		while (reclen) {
+			size_t cplen = MIN(reclen, sizeof(buf));
+
+			rc = eeprom_read(cf->eeprom, addr, buf, cplen);
+			if (rc) {
+				break;
+			}
+			addr += cplen;
+
+			rc = eeprom_write(cf->eeprom, wraddr, buf, cplen);
+			if (rc) {
+				break;
+			}
+			wraddr += cplen;
+			reclen -= cplen;
+		}
```
Maybe this is not a big issue but it seems that compression rewrites data when not needed. For example:
1) lets assume having N bytes of valid records within EEPROM followed by a mix of valid/invalid record, EEPROM is full.
2) we attempt to add/modify record not existing within N first bytes;
3) the compression gets triggered and rewrites first N bytes and then compresses the mix of valid/invalid records.

Laczen replied:
True, but IMO is not important, compression should be called rarely (no need to complicate rarely called routines).

de-nordic reviewed (changes requested):
Some comments provided.

de-nordic commented:
One more thing: I guess that marking records as invalid by spoiling the crc is fine, but for long records it means that we have to calculate crc for every invalid record just to figure out that we need to jump it over; could be for example use the msb (sign bit) of the record length? I know that this would reduce record length to 32k but seems long enough and would reduce needed eeprom reads to figure out where valid records are.

nvlsianpu commented:
^^ a 32 kB record is not a case for this solution anyway.

Laczen commented:
> One more thing: I guess that marking records as invalid by spoiling the crc is fine, but for long records it means that we have to calculate crc for every invalid record just to figure out that we need to jump it over; could be for example use the msb (sign bit) of the record length? I know that this would reduce record length to 32k but seems long enough and would reduce needed eeprom reads to figure out where valid records are.

@de-nordic, thanks for the review. I will comment on your other remarks later. Regarding this remark, yes this could be done, but I don't like it because there would be two methods to identify invalid records.  Eeprom read speed and crc calculation should be fast enough not to do this "optimization".

jharris-intel commented:
>I don't like it because there would be two methods to identify invalid records.  

That's almost a feature.

Using the CRC for deleted record detection, although elegant, makes it very difficult to distinguish between "someone intentionally deleted this record" and "something corrupted this record".

Laczen commented on subsys/settings/src/settings_eeprom.c:
```diff
@@ -0,0 +1,483 @@
+		return -EIO;
+	}
+
+	namelen = strlen(name);
+
+	if (namelen > (SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN)) {
+		return -EINVAL;
+	}
```
I think this is unnecessary "optimization,", the syscall will be made in all non-error cases, what is the use in trying to avoid it in the error case?

de-nordic replied:
> I think this is unnecessary "optimization,", the syscall will be made in all non-error cases, what is the use in trying to avoid it in the error case?

I think it is unnecessary call for non-error cases.

Laczen commented on subsys/settings/src/settings_eeprom.c:
```diff
@@ -0,0 +1,483 @@
+		LOG_DBG("Area to big to fit EEPROM");
+		return -EINVAL;
+	}
+	if (cf->size <= hdrsize) {
+		LOG_DBG("EEPROM to small for settings");
+		return -EIO;
+	}
```
Same comment as above

de-nordic commented:
> 
> @de-nordic, thanks for the review. I will comment on your other remarks later. Regarding this remark, yes this could be done, but I don't like it because there would be two methods to identify invalid records.

No, there will be one method to identify invalid record and one to identify invalid data (https://github.com/zephyrproject-rtos/zephyr/pull/31826#issuecomment-789045332) which also makes record invalid.

>Eeprom read speed and crc calculation should be fast enough not to do this "optimization".

Each EEPROM read may occupy bus, require jumping to syscalls, etc  so yeah this is "optimization"; also: sequential reads require EEPROM circuit to be powered for longer, so in case of aggressive PM this may not be desired.

Laczen commented on subsys/settings/src/settings_eeprom.c:
```diff
@@ -0,0 +1,483 @@
+		 * id if needed
+		 */
+		crc16cal = 0xffff;
+		while ((dataend - *addr) > 0) {
+			size_t rdlen = MIN((dataend - *addr), sizeof(buf));
+
+			rc = eeprom_read(cf->eeprom, *addr, buf, rdlen);
+			if (rc) {
+				break;
+			}
+			/* copy the id if required */
+			if ((id) && ((dataidend - *addr) > 0)) {
+				size_t cplen = MIN((dataidend - *addr), rdlen);
+
+				memcpy(id + (*addr - datastart), buf, cplen);
+			}
+			/* calculate the crc16 */
+			crc16cal = crc16_ccitt(crc16cal, buf, rdlen);
+			*addr += rdlen;
+		}
```
I will have a look at this, it was just to minimize reads.

Laczen commented on subsys/settings/src/settings_eeprom.c:
```diff
@@ -0,0 +1,483 @@
+		}
+	}
+
+	/* Invalidate the entry for a delete or length change */
+	if ((matchdatalen) && (val_len != matchdatalen)) {
+		rc = settings_eeprom_invalidate(cf, matchend);
+		if (rc || (!val_len)) {
+			return rc;
+		}
+	}
```
At the moment you write a new setting you know that the old data is no longer valid. Reloading the old data might be a lot worse than considering the data as deleted. Considering the old data as valid is a possible security breach: getting the device to reboot before updating a security key makes it reload the old key. So IMO it is better to first delete.

Laczen commented on subsys/settings/src/settings_eeprom.c:
```diff
@@ -0,0 +1,483 @@
+	crc16 = crc16_ccitt(crc16, value, val_len);
+
+	/* Overwrite the old record if the length does not change */
+	if ((matchdatalen) && (val_len == matchdatalen)) {
+		addr = matchend - (val_len + sizeof(crc16));
+		rc = eeprom_write(cf->eeprom, addr, value, val_len);
+		if (rc) {
+			return rc;
+		}
+		addr += val_len;
+		return eeprom_write(cf->eeprom, addr, &crc16, sizeof(crc16));
+	}
```
see previous comment

Laczen commented:
@de-nordic, @jharris-intel, there is another reason not to do this: If the update of the record length fails the complete 'fs' might become invalid.

de-nordic commented:
> 
> 
> @de-nordic, @jharris-intel, there is another reason not to do this: If the update of the record length fails the complete 'fs' might become invalid.

Maybe then there is a need for another field that will hold the flag.

Laczen commented:
> > @de-nordic, @jharris-intel, there is another reason not to do this: If the update of the record length fails the complete 'fs' might become invalid.
> 
> Maybe then there is a need for another field that will hold the flag.
> Also if you mention that: when compress gets interrupted while moving data it may already move the length but not the rest of record. When you reload data after such event the incomplete record and everything that follows it is lost.

Another field just to avoid some reads on deleted entries that might not even be there? You must be joking. I also see no reason the distinguish between a deleted item and a bad write.

Laczen commented:
@de-nordic, your comment regarding the problem with interruption during compression is correct and not easily fixed. I am thinking about a solution but have not found a good one yet. Therefore I marked the PR as DNM.

Laczen commented:
> @Laczen Any progress on this?

@henrikbrixandersen No, I have not found a way to avoid corruption when compression is interrupted.

henrikbrixandersen commented:
> @henrikbrixandersen No, I have not found a way to avoid corruption when compression is interrupted.

I suggest moving this PR to draft state until a solution has been found and implemented.

Laczen commented:
@henrikbrixandersen, would it be ok to just close it ?

henrikbrixandersen commented:
> @henrikbrixandersen, would it be ok to just close it ?

That's up to you.

Laczen commented:
I will come back on this when I have a good solution.
